### DISCUSSION TITLE
AMReX STL functionality

### DISCUSSION THREAD

**Discussion by sshivank** [AMReX-Codes/amrex#3221] [Q&A] [answered]
Hello
I was trying to read an STL file using the AMReX EB STL utilities. I prescribe the eb2 geometry type as "stl" in my input script, and I add the STL file name. Then I call the function `EB2::Build(geom, required_level, max_level, 5, true)`. The code runs on a single core. Volume fractions look correct, and AMR levels work as well. But when I change the number of processors, the STL is not correctly recognized. Only a certain portion of the geometry is recognized. Does the STL functionality work with just one core? If not, could you please suggest what I am doing incorrectly? After testing it for my geometry, I tried the Airfoil stl provided in the AMReX Tutorials. That shows the same behavior as well.
![AMReX_STL_airfoil_XY](https://user-images.githubusercontent.com/51785194/228114152-df162348-3bff-4bb5-93bd-88dff270098f.png)

**Comment by hsitaram** [comment, marked as answer]
Hi @sshivank ,

The STL functionality is not perfect yet, but thank you for spotting this error. The problem is that the triangle data is not getting broadcasted across processors, only getting allocated here - https://github.com/AMReX-Codes/amrex/blob/4fab19a02359d05f011b5a5ee4aacfda4b86f640/Src/EB/AMReX_EB_STL_utils.cpp#L243

If the stl read was done by all processors, this should work fine, but that can get memory intense for large stl data. Things work fine If you turn off this line and similar in the read functions - https://github.com/AMReX-Codes/amrex/blob/4fab19a02359d05f011b5a5ee4aacfda4b86f640/Src/EB/AMReX_EB_STL_utils.cpp#L98

Let me see if there is a cleaner way to fix this. @WeiqunZhang - do you know of an easy amrexy way to Bcast GPU-pinnedvector of structs. Bcast(vec.dataPtr, vec.size) didn't work out.

**Reply by WeiqunZhang** [reply]
I think `ParallelDescriptor::Bcast((char*)(vec.dataPtr()), vec.size()*sizeof(Triangle))` should work.

**Reply by WeiqunZhang** [reply]
@hsitaram If that works, could you submit a PR to fix this?

**Reply by hsitaram** [reply]
Yes, that worked great! will submit a PR soon

**Reply by sshivank** [reply]
Hi @hsitaram, for now, turning off the `if (ParallelDescriptor::IOProcessor()) {` lines worked. Thanks for the help.

**Reply by hsitaram** [reply]
@sshivank , we have fixed this issue. Please try again using the latest amrex.

**Reply by sshivank** [reply]
It worked. Thank you.

**Comment by rmrsk** [comment]
@sshivank If you are interested, we have functionality for turning STLs (and other surface meshes) into signed distance fields, which lets you use the STL geometry like any other implicit function. It'll also be faster than native AMReX for larger triangle meshes since it uses a more clever lookup algorithm (AMReX goes linearly through these, I think).
